core/include/drivers/rstctrl.h

```diff
@@ -0,0 +1,228 @@
+ * @ops: Operators of the reset controller
+ * @exclusive: Set when a consumer has exclusive control on the reset level
+ */
+struct rstctrl {
+	const struct rstctrl_ops *ops;
+	bool exclusive;
+};
+
+/**
+ * RSTCTRL_DECLARE - Declare a reset controller driver with a single
+ * DTT compatible string.
+ *
+ * @__name: Reset controller driver name
+ * @__compat: Compatible string
+ * @__probe: Reset controller probe function
+ */
+#define RSTCTRL_DT_DECLARE(__name, __compat, __probe) \
+	static const struct dt_device_match __name ## _match_table[] = { \
+		{ .compatible = __compat }, \
+		{ } \
+	}; \
+	const struct dt_driver __name ## _dt_driver __dt_driver = { \
+		.name = # __name, \
+		.type = DT_DRIVER_RSTCTRL, \
+		.match_table = __name ## _match_table, \
+		.probe = __probe, \
+	}
+
+/*
+ * Platform driver may ignore the timeout hint according to their
+ * capabilities. RSTCTRL_NO_TIMEOUT specifies no timeout hint.
+ */
+#define RSTCTRL_NO_TIMEOUT	0
+
+/*
+ * rstctrl_assert_to - Assert reset control possibly with timeout
+ * rstctrl_assert - Assert reset control
+ * rstctrl_deassert_to - Deassert reset control possibly with timeout
+ * rstctrl_deassert - Deassert reset control
+ *
+ * @rstctrl: Reset controller
+ * @to_us: Timeout in microseconds
+ * Return a TEE_Result compliant code
+ */
+static inline TEE_Result rstctrl_assert_to(struct rstctrl *rstctrl,
+					   unsigned int to_us)
+{
+	return rstctrl->ops->assert_level(rstctrl, to_us);
+}
+
+static inline TEE_Result rstctrl_assert(struct rstctrl *rstctrl)
+{
+	return rstctrl_assert_to(rstctrl, RSTCTRL_NO_TIMEOUT);
+}
+
+static inline TEE_Result rstctrl_deassert_to(struct rstctrl *rstctrl,
+					     unsigned int to_us)
+{
+	return rstctrl->ops->deassert_level(rstctrl, to_us);
+}
+
+static inline TEE_Result rstctrl_deassert(struct rstctrl *rstctrl)
+{
+	return rstctrl_deassert_to(rstctrl, RSTCTRL_NO_TIMEOUT);
+}
+
+/*
+ * rstctrl_name - Get a name for the reset level control or NULL
+ *
+ * @rstctrl: Reset controller
+ * Return a pointer to  anem or NULL
+ */
+static inline const char *rstctrl_name(struct rstctrl *rstctrl)
+{
+	if (rstctrl->ops->get_name)
+		return rstctrl->ops->get_name(rstctrl);
+
+	return NULL;
+}
+
+/**
+ * rstctrl_dt_get_exclusive - Get exclusive control or target
+ *
+ * @rstctrl: Reference to reset level control
+ * Return a TEE_Result compliant value
+ */
+TEE_Result rstctrl_get_exclusive(struct rstctrl *rstctrl);
+
+/**
+ * rstctrl_put_exclusive - Release exclusive access to target
+ *
+ * @rstctrl: Reference to reset level control
+ */
+void rstctrl_put_exclusive(struct rstctrl *rstctrl);
+
+/**
+ * rstctrl_ops_is_valid - Check reset controller ops is valid
+ *
+ * @ops: Reference to reset controller operator instance
+ */
+static inline bool rstctrl_ops_is_valid(const struct rstctrl_ops *ops)
+{
+	return ops && ops->assert_level && ops->deassert_level;
+}
+
+/**
+ * rstctrl_dt_get_by_idx - Get a reset controller at a specific index in
```

Comment:
I want to rename this to: `dt_get_rstctrl_by_index()` with get `TEE_result` return value:
```c
TEE_Result dt_get_rstctrl_by_index(const void *fdt, int nodeoffset,
				   unsigned int index,
				   struct rstctrl **rstctrl);
```
It will look better in caller's instruction block:
```c
	struct rstctrl *rstctrl_ref = NULL;

	res = dt_get_rstctrl_by_index(fdt, node, index, &rstctrl_ref);
	if (res)
		return res;
	res = rstctrl_get_exclisive(rstctrl_ref);
	if (res)
		return res;
	(...)
```

I'll propose the same for clk: 
```c
TEE_Result dt_get_clk_by_index(const void *fdt, int nodeoffset,
			       unsigned int index, struct clk **clk);
TEE_Result dt_get_clk_by_name(const void *fdt, int nodeoffset,
			      chanr *name, struct clk **clk);
```

Reply:
Looks good to me for the clock functions.